features/cryptocell/FEATURE_CRYPTOCELL310/cmac_alt.c

```diff
@@ -0,0 +1,333 @@
+            mbedtls_free( cmac_ctx );
+            return( MBEDTLS_ERR_PLATFORM_FEATURE_UNSUPPORTED );
+        }
+        default:
+        {
+            mbedtls_free( cmac_ctx );
+            return( MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA );
+        }
+    }
+
+    ctx->cmac_ctx = cmac_ctx;
+    return 0;
+}
+
+static int init_cc( mbedtls_cmac_context_t *cmac_ctx )
+{
+    int ret = 0;
+    SaSiAesUserKeyData_t CC_KeyData;
+    if( SaSi_AesInit( &cmac_ctx->CC_Context, SASI_AES_ENCRYPT,
+                      SASI_AES_MODE_CMAC, SASI_AES_PADDING_NONE) != 0 )
+    {
+        return( MBEDTLS_ERR_PLATFORM_HW_ACCEL_FAILED );
+    }
+
+    CC_KeyData.pKey = cmac_ctx->CC_Key;
+    CC_KeyData.keySize = cmac_ctx->CC_keySizeInBytes;
+
+    if( SaSi_AesSetKey( &cmac_ctx->CC_Context, SASI_AES_USER_KEY,
+                        &CC_KeyData, sizeof( CC_KeyData ) ) != 0 )
+    {
+        ret =  MBEDTLS_ERR_PLATFORM_HW_ACCEL_FAILED;
+        goto exit;
+    }
+
+    cmac_ctx->is_cc_initiated = 1;
+
+exit:
+    return( ret );
+    
+}
+
+int mbedtls_cipher_cmac_update( mbedtls_cipher_context_t *ctx,
+                                const unsigned char *input, size_t ilen )
+{
+    mbedtls_cmac_context_t* cmac_ctx;
+    int ret = 0;
+    size_t block_size;
+
+    if( ctx == NULL || ctx->cipher_info == NULL || input == NULL ||
+        ctx->cmac_ctx == NULL )
+        return( MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA );
+
+    if( ctx == NULL || ctx->cipher_info == NULL )
+        return( MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA );
+
+
+    block_size = ctx->cipher_info->block_size;
+    if( block_size != SASI_AES_BLOCK_SIZE_IN_BYTES )
+        return( MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA );
+
+    cmac_ctx = ctx->cmac_ctx;
+
+    if( cmac_ctx->is_cc_initiated == 0 )
+    {
+        ret = init_cc( cmac_ctx );
+        if( ret != 0 )
+            goto exit;
+    }
+
+    /* Is there data still to process from the last call?
+     */
+    if( cmac_ctx->unprocessed_len > 0  )
+    {
+        const size_t size_to_copy = ilen > ( block_size - cmac_ctx->unprocessed_len ) ?
+                                    block_size - cmac_ctx->unprocessed_len : ilen;
+        memcpy( &cmac_ctx->unprocessed_block[cmac_ctx->unprocessed_len],
+                input, size_to_copy );
+        cmac_ctx->unprocessed_len += size_to_copy;
+        input += size_to_copy;
+        ilen -= size_to_copy;
+
+        /*
+         * Process the unproccessed data, in case it reached a full AES block,
+         * and there is there is still input data.
+         */
+        if( cmac_ctx->unprocessed_len == SASI_AES_BLOCK_SIZE_IN_BYTES && ilen > 0 )
+        {
+            if( SaSi_AesBlock( &cmac_ctx->CC_Context, cmac_ctx->unprocessed_block,
+                              SASI_AES_BLOCK_SIZE_IN_BYTES, NULL ) != 0 )
+            {
+                ret = MBEDTLS_ERR_PLATFORM_HW_ACCEL_FAILED;
+                goto exit;
+            }
+            cmac_ctx->unprocessed_len = 0;
+        }
+    }
+
+    if( ilen > 0 )
```

Comment:
I think the assumption for the code in this if statement is that there are is no data in unprocessed_block if `ilen == 0`. Could we write that down in a comment? I just think that the precondition is not obvious.

Reply:
But this is not correct.
It is possible to have `ilen == 0` and data in `unprocessed_block`. This is in case the function is given an empty in buffer. However, in this case, if `unprocessed_block` is not full, we don't do anything with it, until `mbedtls_cipher_cmac_finish()` is called. 

In addition, it is possible that a non empty buffer is given as input, but it is very short ( e.g. 1 byte ), and the `unprocessed_block has not reached 16 bytes after storing the byte.

Reply:
Sorry, my comment was incorrect, I meant to say that the code in the if statement is executed only when unprocessed_block doesnt have any data.